lib/components/artworks/index.js

```diff
@@ -0,0 +1,224 @@
+//    won't be too different between artworks.
+// 3. Get artwork heights by either:
+//    - calculating the item size upfront with aspect ratio and a static height for the text labels.
+//    - leting the artwork component do a layout pass and calculate its own height based on the column width.
+// 4. Update height of grid to encompass all items.
+
+import Relay from 'react-relay';
+import React from 'react-native';
+const { View, ScrollView, StyleSheet } = React;
+
+import Artwork from './artwork';
+import { metaphysicsURL } from '../../relay/config';
+
+/**
+ * TODOs:
+ * - currently all the code assumes column layout
+ *   - do no invert aspect ratios in row layout
+ * - deal with edge-cases when calculating in which section an artwork should go
+ *   - see ARMasonryCollectionViewLayout for details on how to deal with last works sticking out
+ *   - the calculation currently only takes into account the size of the image, not if e.g. the sale message is present
+ */
+class Artworks extends React.Component {
+  static defaultProps = {
+    sectionDirection: 'column',
+    sectionCount: 2,
+    sectionMargin: 8,
+    itemMargin: 8,
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = { sectionDimension: 0, artworks: this.props.artworks, page: 1 };
+
+    this.onLayout = this.onLayout.bind(this);
+    this.onScroll = this.onScroll.bind(this);
+
+    this._sentEndForContentLength = null;
+  }
+
+  fetchNextPage() {
+    const nextPage = this.state.page + 1;
+    const query = artworksQuery(this.props.relay.route.params.artistID, nextPage);
+    fetch(metaphysicsURL, { method: 'post',
+                            headers: { 'Content-Type': 'application/json' },
+                            body: JSON.stringify({ query }) })
+      .then((response) => {
+        if (response.status >= 200 && response.status < 300) {
+          return response
+        } else {
+          const error = new Error(response.statusText)
+          error.response = response
+          throw error
+        }
+      })
+      .then((response) => response.json())
+      .then(({ data }) => {
+        const artworks = data.artist.artworks;
+        this.setState({ page: nextPage, artworks: this.state.artworks.concat(artworks) });
+      })
+      .catch((error) => {
+        console.log(error);
+      });
+  }
+
+  onLayout(event: LayoutEvent) {
+    const layout = event.nativeEvent.layout;
+    if (layout.width > 0) {
+      // This is the sum of all margins in between sections, so do not count to the right of last column.
+      const sectionMargins = this.props.sectionMargin * (this.props.sectionCount - 1);
+      this.setState({ sectionDimension: (layout.width - sectionMargins) / this.props.sectionCount });
+    }
+  }
+
+  sectionedArtworks() {
+    const sectionedArtworks = [];
+    const sectionRatioSums = [];
+    for (let i = 0; i < this.props.sectionCount; i++) {
```

Comment:
This isn’t an array.

I could definitely have used `forEach` on the artworks array below. I guess at some point I was using the `i` variable for something else as well.